packages/user/src/model.ts

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { IStateDB } from '@jupyterlab/statedb';
+import { ISignal, Signal } from '@lumino/signaling';
+import { JSONObject, ReadonlyPartialJSONObject, UUID } from '@lumino/coreutils';
+import * as env from 'lib0/environment';
+
+import { ICurrentUser, IUser, USER } from './tokens';
+import { getAnonymousUserName, getRandomColor } from './utils';
+
+/**
+ * Default user implementation.
+ */
+export class User implements ICurrentUser {
+  private _id: string;
+  private _name: string;
+  private _username: string;
+  private _color: string;
+  private _anonymous: boolean;
+  private _role: IUser.ROLE;
+  private _cursor?: IUser.Cursor;
+
+  private _isReady = false;
+  private _state: IStateDB;
+  private _ready = new Signal<User, boolean>(this);
+  private _changed = new Signal<User, void>(this);
+
+  /**
+   * Constructor of the User class.
+   *
+   * @argument state: IStateDB to store users information.
+   */
+  constructor(state: IStateDB) {
+    this._state = state;
+    this._fetchUser()
+      .then(() => {
+        this._isReady = true;
+        this._ready.emit(true);
+      })
+      .catch(e => console.error(e));
+  }
+
+  /**
+   * User's ID.
+   */
+  get id(): string {
+    return this._id;
+  }
+
+  /**
+   * User's name.
+   */
+  get name(): string {
+    return this._name;
+  }
+
+  /**
+   * User's Username.
+   */
+  get username(): string {
+    return this._username;
+  }
+
+  /**
+   * User's cursor color and icon color if avatar_url is undefined
+   * (there is no image).
+   */
+  get color(): string {
+    return this._color;
+  }
+
+  /**
+   * Whether the user is anonymous or not.
+   *
+   * NOTE: Jupyter server doesn't handle user's identity so, by default every user
+   * is anonymous unless a third-party extension provides the ICurrentUser token retrieving
+   * the user identity from a third-party identity provider as GitHub, Google, etc.
+   */
+  get anonymous(): boolean {
+    return this._anonymous;
+  }
+
+  /**
+   * User's role.
+   *
+   * NOTE: Jupyter Server and JupyterLab doesn't implement a role-base access control (RBAC) yet.
+   * This attribute is here to start introducing RBAC to JupyterLab's interface. At the moment every
+   * user has the role ADMIN.
+   */
+  get role(): IUser.ROLE {
```

Comment:
While I think it is the right direction to start adding roles, I don't know that this is the right place to put this. The reason is that there is little in place to ensure that the role returned here is the authoritative role given to the user by the system. For, example, an attacking extension could disable `ICurrentUser` and provide its own implementation that escalates the role. Thoughts on removing for now until we work through those details?

Reply:
Thanks for reviewing the PR @ellisonbg.

You are right. This is not the right place to ensure the authorization. All the authentication and authorization have to be handled by the backend by adding RBAC to all the end-points to return `401` when the user doesn't have the proper permissions.

Nevertheless, we need to somehow show different UI components depending on the user's role. For example, a user with `read` permission should not see the notebook in edit mode. And sure, any extension would be able to change the ICurrentUser plugin and assign any role to the user, but that doesn't mean that the user can edit the document. When the user tries to edit the document, the backed will respond with a `401`.

Reply:
The other reason to start adding the role already is that at the moment, all the users are `admin` (have the higher role). If we start to modify the UI deactivating buttons for other roles (for example, a user with role `read` can not see the button to open settings). This won't affect the actual behavior of JupyterLab because every user is `admin`.

This property doesn't alter the frontend but allows us to start adding RBAC to the front end without waiting until the backend is done.

Reply:
Coming from JupyterHub, the main use of roles is to resolve to permissions (scopes). All permission checks are then direct scope checks, and roles are purely a point of storage, resolved immediately to scopes and never used directly thereafter. In an environment like JupyterHub, JupyterLab (client-side) wouldn't have any use for roles because it is the responsibility of the server to map request credentials to a set of scopes. _But_ duplicating permission scope checks server & client side might be a maintenance pain!

But if your definition of 'role' is a little different (i.e. they are a few fixed bundles of permissions, not arbitrary named sets of permissions), then each JupyterLab 'role' would map just fine to a single JupyterHub 'scope', and that's no problem.

Reply:
Hi @minrk, thank you soo much for the feedback.

First of all, it would be great if we can all set a meeting to sync between JupyterHub, Jupyter-Server and JupyterLab to discuss and decide how to proceed with the implementation of RBAC across projects. I suggest to use the RTC meeting every two weeks.

You (JupyterHub team) have done an amazing job in 2.0 and I think we should follow the same pattern.

> In an environment like JupyterHub, JupyterLab (client-side) wouldn't have any use for roles because it is the responsibility of the server to map request credentials to a set of scopes

But what about fine grade permission to the content, and kernels API on Jupyter Server? For what I understood from [the documentation](https://jupyterhub.readthedocs.io/en/rbac/rbac/index.html) correct me if I'm worng, JupyterHub assign permissions to manage servers, but what about allowing multiple users to access a server with different permissions? For example, both can write documents but only one can send a execute request to the kernel.

Reply:
> I suggest to use the RTC meeting every two weeks.

Yes, that makes sense. I will try to come once in a while, but that meeting is at a very difficult time for me, so I cannot come on a regular basis. I think I can come next week, though.

> But what about fine grade permission to the content, and kernels API on Jupyter Server? 

For JupyterHub, I believe that should be implemented as custom scopes for JupyterHub, which 2.0 doesn't _yet_ support. But I plan to, once we figure out more about what it should look like here. Part of my goal for this collaboration is to establish:

1. what do scopes look like for Jupyter Server, and
2. how can they be represented in JupyterHub

I see two ways to do it - express the full fine-grained scopes in JupyterHub, _or_ express only a short list of roles in JupyterHub, and then map them to permissions in Jupyter serer. The latter case only really works if single-server roles are relatively static, though.

But for designing permissions in JupyterLab, one thing you should assume is the ability to store your own scopes and role assignments in JupyterHub. Even if that feature doesn't exist yet.

Reply:
> I will try to come once in a while, but that meeting is at a very difficult time for me, so I cannot come on a regular basis. I think I can come next week, though.

We changed the time since it was late for Europe. Now it is Monday 5:00 pm CET (the next one is next week, Monday 13th)

> But I plan to, once we figure out more about what it should look like here

Then we should work together to come out with the shape of RBAC in JupyteLab.

Pining @davidbrochart since he might be interested in the discussion.

Reply:
> Now it is Monday 5:00 pm CET (the next one is next week, Monday 13th)

Yeah, 5-7 CET will always be hard, every day for me. But I can move things around once in a while. There's no way around it, I just can't participate regularly in synchronous meetings with the US (I think it's a good practice for all community projects to minimize synchronized meetings).

FWIW, over in jupyter-server, there are no plans to implement roles (https://github.com/jupyter-server/jupyter_server/pull/165 and https://github.com/jupyter-server/jupyter_server/pull/540) only scopes. That's all that's needed for fine-grained permissions. Roles are a convenient way for managing assignment of scopes to users, but not part of the actual enforcement of permissions.

So if there is something to store here, I would probably say it should be `scopes` or `permissions` rather than `role`. And it should be a collection, not a single value (that would be true for roles, too, btw). That could be a dict of `{resourceName: permissionLevel}` (probably easiest to work with, matches current [plan](https://github.com/jupyter-server/jupyter_server/pull/165) in jupyter-server), or a list of string scopes.

> Then we should work together to come out with the shape of RBAC in JupyterLab.

That's my plan! As much as possible via GitHub issues, to maximize visibility, participation, etc.

Reply:
I'm going to remove the property `role` from the `ICurrentUser` interface, this way we can merge the PR.
Meanwhile, I will look into the Jupyter Server and open another PR for RBAC later.
For now, there is an issue open to discuss how to implement RBAC in the frontend and I think we should continue the conversation there:
* https://github.com/jupyterlab/jupyterlab/issues/11355